src/rust/ide/enso-protocol/src/project_manager.rs

```diff
@@ -0,0 +1,218 @@
+
+        let delete_result = mock_client.delete_project(expected_uuid.clone());
+        result(&mut fixture, delete_result).expect_err("Project shouldn't exist.");
+
+        let uuid = mock_client.create_project("HelloWorld".into());
+        let uuid = result(&mut fixture,uuid).expect("Couldn't create project");
+        assert_eq!(uuid, expected_uuid);
+
+        let close_result = result(&mut fixture,mock_client.close_project(uuid.clone()));
+        close_result.expect_err("Project shouldn't be open.");
+
+        let ip_with_socket = result(&mut fixture,mock_client.open_project(uuid.clone()));
+        let ip_with_socket = ip_with_socket.expect("Couldn't open project");
+        assert_eq!(ip_with_socket, expected_ip_with_socket);
+
+        mock_client.set_close_project_result(expected_uuid.clone(), Ok(()));
+        result(&mut fixture,mock_client.close_project(uuid)).expect("Couldn't close project.");
+
+        mock_client.set_delete_project_result(expected_uuid.clone(), Ok(()));
+        result(&mut fixture,mock_client.delete_project(uuid)).expect("Couldn't delete project.");
+    }
+
+    #[test]
+    fn list_projects() {
+        let mut fixture     = setup();
+        let mut mock_client = fixture.client.clone();
+        let project1        = ProjectMetaData {
+            name        : "project1".to_string(),
+            id          : Uuid::default(),
+            last_opened : chrono::DateTime::parse_from_rfc3339("2020-01-07T21:25:26Z").unwrap()
+        };
+        let project2 = ProjectMetaData {
+            name        : "project2".to_string(),
+            id          : Uuid::default(),
+            last_opened : chrono::DateTime::parse_from_rfc3339("2020-02-02T13:15:20Z").unwrap()
+        };
+        let expected_recent_projects = vec![project1,project2];
+        let sample1                  = ProjectMetaData {
+            name        : "sample1".to_string(),
+            id          : Uuid::default(),
+            last_opened : chrono::DateTime::parse_from_rfc3339("2019-11-23T05:30:12Z").unwrap()
+        };
+        let sample2 = ProjectMetaData {
+            name        : "sample2".to_string(),
+            id          : Uuid::default(),
+            last_opened : chrono::DateTime::parse_from_rfc3339("2019-12-25T00:10:58Z").unwrap()
+        };
+        let expected_sample_projects = vec![sample1,sample2];
+        mock_client.set_list_recent_result(2,Ok(expected_recent_projects.clone()));
+        mock_client.set_list_sample_result(2,Ok(expected_sample_projects.clone()));
+
+        let recent_projects = result(&mut fixture,mock_client.list_recent(2)).expect("Couldn't get recent projects.");
+        assert_eq!(recent_projects, expected_recent_projects);
+        let sample_projects = result(&mut fixture,mock_client.list_sample(2)).expect("Couldn't get sample projects.");
+        assert_eq!(sample_projects, expected_sample_projects);
+    }
+}
```

Comment:
These tests looks like testing mocks.

We should also have a tests which connect with the real Project Management Service. It could be test which have `#[ignore]` flag so CI will not fail, but when someone would like to run test with the real Engine running he can comment this flag. This is how some tests of parser works (see ide/parser/tests.rs).

So please add `tests` directory in your crate and write some simple scenarios using real Client (like list projects, create new one, open and close it) and then run it with Engine process working and see if everything works (and if you see reaction in Enso logs).

When you find out how to run engine services, please add the information about these tests in our doc/ directory with instruction how to run them (with information how to run engine or a reference to such instruction if exists)

Reply:
What @farmaazon writes about are integration tests. 
What I want and find way more important are unit tests. Each client method should be tested, in particular that:
a) once call is made, an expected JSON message is emitted;
b) once a predefined answer JSON is received by the client, the Future yields expected value.
Use MockTransport for this.